pyrs990/pyrs990/index.py

```diff
@@ -1,180 +1,159 @@
-        self._filters = filters
-        self._records = records
+    _annual_years: Set[AnnualYear] = set()
 
-    def __iter__(self) -> Iterator[IndexRecord]:
-        for record in self._records:
-            passed = True
-            for cb in self._filters:
-                if not cb(record):
-                    passed = False
-                    break
-            if passed:
-                yield record
+    _bmf_indices: List[BMFIndex] = []
 
-    def augment(self, other: Iterable[IndexRecord]) -> Index:
-        return AugmentedIndex(self, other)
+    _bmf_regions: Set[BMFRegion] = set()
 
-    def filter(self, cb: FilterCallback[IndexRecord]) -> Index:
-        return FilteredIndex(self, cb)
+    _filters: Set[IndexFilter] = set()
 
+    _length: Optional[int] = None
 
-class AugmentedIndex(Index):
-    """
-    An `Index` that can combine the records from several different indices
-    into one logical index.
-
-    TODO: Just hold the iterables and iterate them sequentially
-    """
-
-    def __init__(self, original: Index, others: Iterable[IndexRecord]):
-        records = list(original._records) + list(others)
-        super().__init__(records)
-
-
-class FilteredIndex(Index):
-    """
-    An `Index` that can be created from a parent and filter rather than a
-    file and collection of filters. This is what is actually returned from
-    the `filter()` method.
-    """
-
-    def __init__(
-        self, parent: Index, cb: FilterCallback[IndexRecord],
-    ):
-        filters = list(parent._filters) + [cb]
-
-        # We share the list of records since it could be quite large. This
-        # shouldn't prevent concurrent iteration since we ultimately delegate
-        # to the list's iterator.
-        records = parent._records
-
-        super().__init__(records, filters)
+    def __iter__(self) -> Iterator[IndexRecord]:
+        # TODO: Be smarter about which one to iterate first based on filters
+        # TODO: Consider whether re-instantiating all these tuples is too slow
+        for annual_index in self._annual_indices:
+            for bmf_index in self._bmf_indices:
+                for annual_record in annual_index.values():
+                    ein = EIN(annual_record.ein)
+                    if ein not in bmf_index:
+                        _logger.info(f"EIN not found in BMF index: {ein}")
+                        continue
+
+                    bmf_record = bmf_index[ein]
+
+                    index_record = IndexRecord(annual_record, bmf_record)
+                    passed = True
+                    for cb in self._filters:
+                        if not cb(index_record):
+                            passed = False
+                            break
+
+                    if passed:
+                        yield index_record
+
+    def __len__(self) -> int:
+        if self._length is not None:
+            return self._length
+
+        # TODO: This is stupid, don't instantiate the tuples, just count
```

Comment:
.......What tuples?

Reply:
The records themselves, we don't need to create them to count how many there are, we can just count how many things that would normally be turned into records there are. Like if you want to know how many bread sandwiches you can make you can just count the slices of break, you don't have to actually make sandwiches and then count them.